pandas/_libs/tslibs/timezones.pyx

```diff
@@ -275,3 +275,84 @@ cdef object get_dst_info(object tz):
         dst_cache[cache_key] = (trans, deltas, typ)
 
     return dst_cache[cache_key]
+
+
+def _infer_tzinfo(start, end):
+    def _infer(a, b):
+        tz = a.tzinfo
+        if b and b.tzinfo:
+            if not (get_timezone(tz) == get_timezone(b.tzinfo)):
+                raise AssertionError('Inputs must both have the same timezone,'
+                                     ' {timezone1} != {timezone2}'
+                                     .format(timezone1=tz, timezone2=b.tzinfo))
+        return tz
+
+    tz = None
+    if start is not None:
+        tz = _infer(start, end)
+    elif end is not None:
+        tz = _infer(end, start)
+    return tz
+
+
+cdef ndarray[int64_t] _infer_dst(ndarray[int64_t] vals,
```

Comment:
you are removing *part* of this function. I think its better to move almost *all* of it  to timezones or leave it. (ok with latter for now, maybe do former at some point). Otherwise you end up splitting the logic / comments in 2 places.

I think this requires a bit more thought here.

Reply:
This is the appropriate block to separate out.  It is exclusively focused on getting DST info for use in the original function.  The rest of the function is thematically more related to _TSObject conversion.

In particular, `pandas_datetimestruct` doesn't belong anywhere near `timezones`.